national_rail/transform.py

```diff
@@ -0,0 +1,163 @@
+
+    return datetime.fromisoformat(time.replace('Z', ''))
+
+
+def check_creation_within_last_5_minutes(creation_time: datetime) -> bool:
+    """ Checks that the creation time of an incident occurred within the last 5 minutes. """
+
+    current_time = datetime.now()
+    time_diff = current_time - creation_time
+
+    return time_diff <= timedelta(minutes=5)
+
+
+def process_pt_incidents(incidents: list[ET.Element], namespaces: dict) -> list[dict]:
+    """ Extracts relevant data for each incident reported. """
+    dataset = []
+
+    for incident in incidents:
+
+        creation_time = convert_to_datetime(find_text_element(
+            incident, 'ns:CreationTime', namespaces))
+
+        if not check_creation_within_last_5_minutes(creation_time):
+            # remove not to allow all incidents
+            break
+
+        incident_number = find_text_element(
+            incident, 'ns:IncidentNumber', namespaces)
+
+        operator_codes = find_all_text_elements(
+            incident, "ns:Affects/ns:Operators/ns:AffectedOperator/ns:OperatorRef", namespaces)
+
+        start_time = find_text_element(
+            incident, 'ns:ValidityPeriod/com:StartTime', namespaces)
+        end_time = find_text_element(
+            incident, 'ns:ValidityPeriod/com:EndTime', namespaces)
+
+        is_planned = find_text_element(incident, 'ns:Planned', namespaces)
+
+        summary = find_text_element(incident, 'ns:Summary', namespaces)
+
+        description = convert_html_to_text(find_text_element(
+            incident, 'ns:Description', namespaces))
+
+        uri = find_text_element(
+            incident, "ns:InfoLinks/ns:InfoLink/ns:Uri", namespaces).replace('/n', " ").strip()
+
+        routes_affected = convert_html_to_text(find_text_element(
+            incident, "ns:Affects/ns:RoutesAffected", namespaces))
+
+        data = {
+            'incident_number': incident_number,
+            'operator_codes': operator_codes,  # list
+            'creation_time': creation_time,
+            'start_time': start_time,
+            'end_time': end_time,
+            'is_planned': is_planned,
+            'summary': summary,
+            'description': description,
+            'uri': uri,
+            'routes_affected': routes_affected
+        }
+        dataset.append(data)
+
+        print(incident_number)
+        print(operator_codes)
+        print(creation_time)
+        print(start_time, end_time)
+        print(is_planned)
+        print(summary)
+        print(uri)
+        print(routes_affected)
+        print(description)
+
+        print("\n")
+
+    return dataset
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.INFO,
+                        format="%(asctime)s - %(levelname)s - %(message)s")
+
+    nr_namespaces = {'ns': 'http://nationalrail.co.uk/xml/incident',
+                     'com': 'http://nationalrail.co.uk/xml/common'}
+
```

Comment:
maybe have a logging info to mention that transformation has started?

Reply:
Yes good call